include/boost/geometry/util/promote_integral.hpp

```diff
@@ -0,0 +1,276 @@
+>
+class promote_integral
+{
+private:
+    static bool const is_unsigned = boost::is_unsigned<T>::type::value;
+
+#if !defined(BOOST_GEOMETRY_NO_MULTIPRECISION_INTEGER)
+    // Define the proper check policy for the multiprecision integer
+    typedef typename boost::mpl::if_c
+        <
+            UseCheckedMultiprecisionInteger,
+            boost::integral_constant
+                <
+                    boost::multiprecision::cpp_int_check_type,
+                    boost::multiprecision::checked
+                >,
+            boost::integral_constant
+                <
+                    boost::multiprecision::cpp_int_check_type,
+                    boost::multiprecision::unchecked
+                >
+        >::type check_policy_type;
+
+    // Meta-function to get the multiprecision integer type for the
+    // given size and type (signed/unsigned)
+    template
+    <
+        unsigned int Size,
+        boost::multiprecision::cpp_integer_type IntegerType
+    >
+    struct multiprecision_integer_type
+    {
+        typedef boost::multiprecision::number
+            <
+                boost::multiprecision::cpp_int_backend
+                    <
+                        Size,
+                        Size,
+                        IntegerType,
+                        check_policy_type::value,
+                        void
+                    >
+            > type;
+    };
+
+    // Meta-function to get the multiprecision signed integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_signed_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::signed_magnitude
+            >::type type;
+    };
+
+    // Meta-function to get the multiprecision unsigned integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_unsigned_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::unsigned_magnitude
+            >::type type;
+    };
+
+    // Define the minimum (and maximum) bit size for the
+    // multiprecision integer needed
+    // If T is the input type and P the promoted type, then the
+    // minimum number of bits for P are (below b stands for the number
+    // of bits of T):
+    // * if T is unsigned and P is unsigned: 2 * b
+    // * if T is signed and P is signed: 2 * b - 1
+    // * if T is unsigned and P is signed: 2 * b + 1
+    typedef typename boost::mpl::if_c
+        <
+            (PromoteUnsignedToUnsigned && is_unsigned),
+            boost::integral_constant
+                <
+                    unsigned int, (2 * CHAR_BIT * sizeof(T))
+                >,
+            boost::integral_constant
+                <
+                    unsigned int,
+                    (2 * CHAR_BIT * sizeof(T) + (is_unsigned ? 1 : -1))
```

Comment:
* I mainly suggested that to avoid warnings about conversion from size_t to unsigned on x64 because sizeof() returns std::size_t (can those warnings be generated also for compile-time constants?). Anyway maybe size_t should then be used here and then converted when it's passed into the cpp_int_backend?
* ok great!
* My point was that in the code the number of bits is decreased for signed promoted to signed (`(is_unsigned ? 1 : -1)`) which seems to be wrong. The number of bits should probably be exactly 2*bits also for signed, not 2*bits-1. Or am I missing something?
* Yes, that's reasonable assumption. Though I have a feeling that some things can be left for the user, e.g. in the examples where Min==Max and void must be passed as the Allocator, Max is always set to the power of 2. My guess is that this should be the fastest. As said, we can postpone it.